platforms/software/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java

```diff
@@ -90,66 +96,81 @@ public enum PathBase {
         PROJECT, GRADLE_USER_HOME
     }
 
-    private Object scriptFile = WrapperDefaults.SCRIPT_PATH;
-    private Object jarFile = WrapperDefaults.JAR_FILE_PATH;
-    private String distributionPath = DEFAULT_DISTRIBUTION_PARENT_NAME;
-    private PathBase distributionBase = WrapperDefaults.DISTRIBUTION_BASE;
-    private String distributionUrl;
-    private String distributionSha256Sum;
     private final GradleVersionResolver gradleVersionResolver = new GradleVersionResolver();
-    private DistributionType distributionType = WrapperDefaults.DISTRIBUTION_TYPE;
-    private String archivePath = WrapperDefaults.ARCHIVE_PATH;
-    private PathBase archiveBase = WrapperDefaults.ARCHIVE_BASE;
     private final Property<Integer> networkTimeout = getProject().getObjects().property(Integer.class);
-    private boolean distributionUrlConfigured = false;
     private final boolean isOffline = getProject().getGradle().getStartParameter().isOffline();
+    private final RegularFileProperty batchScript;
+    private final RegularFileProperty propertiesFile;
 
     public Wrapper() {
         getValidateDistributionUrl().convention(WrapperDefaults.VALIDATE_DISTRIBUTION_URL);
+        getScriptFile().convention(getProject().getObjects().fileProperty()).fileValue(getFileOperations().file(WrapperDefaults.SCRIPT_PATH));
+        getJarFile().convention(getProject().getObjects().fileProperty()).fileValue(getFileOperations().file(WrapperDefaults.JAR_FILE_PATH));
+        getDistributionPath().convention(DEFAULT_DISTRIBUTION_PARENT_NAME);
+        getDistributionBase().convention(WrapperDefaults.DISTRIBUTION_BASE);
+        getDistributionType().convention(WrapperDefaults.DISTRIBUTION_TYPE);
+        getArchivePath().convention(WrapperDefaults.ARCHIVE_PATH);
+        getArchiveBase().convention(WrapperDefaults.ARCHIVE_BASE);
+        getGradleVersion().convention(GradleVersion.current().getVersion());
+        getDistributionUrl().convention(getGradleVersion().zip(getDistributionType(), this::getDistributionUrl));
+        this.batchScript = getObjectFactory().fileProperty().fileProvider(getScriptFile().getLocationOnly().map(file -> WrapperGenerator.getBatchScript(file.getAsFile())));
+        this.propertiesFile = getObjectFactory().fileProperty().fileProvider(getJarFile().getLocationOnly().map(file -> WrapperGenerator.getPropertiesFile(file.getAsFile())));
```

Comment:
💭 Could every convention use the same `getObjectFactory()` vs. `getProject().getObjects()` here?